components/ledger/pkg/storage/sqlstorage/worker/worker.go

```diff
@@ -2,110 +2,137 @@ package worker
-
-	models     []Model[MODEL]
-	modelsChan chan Model[MODEL]
-
-	workerJob WorkerJob[MODEL]
+	pending      []modelsHolder[MODEL]
+	writeChannel chan modelsHolder[MODEL]
+	jobs         chan []modelsHolder[MODEL]
+	releasedJob  chan struct{}
+	workerJob    Job[MODEL]
+	stopChan     chan chan struct{}
 }
 
-func NewWorker[MODEL any](batchSize int, batchTime time.Duration, workerJob WorkerJob[MODEL]) *Worker[MODEL] {
+func NewWorker[MODEL any](workerJob Job[MODEL]) *Worker[MODEL] {
 	return &Worker[MODEL]{
-		batchSize: batchSize,
-		batchTime: batchTime,
-		workerJob: workerJob,
-
-		models:     nil,
-		modelsChan: make(chan Model[MODEL], 1024),
+		workerJob:    workerJob,
+		pending:      make([]modelsHolder[MODEL], 0), // TODO(gfyrag): we need to limit the worker capacity
+		jobs:         make(chan []modelsHolder[MODEL]),
+		writeChannel: make(chan modelsHolder[MODEL], 1024), // TODO(gfyrag): Make configurable
+		releasedJob:  make(chan struct{}, 1),
+		stopChan:     make(chan chan struct{}, 1),
 	}
 }
 
-// Run should be called in a goroutine
-func (w *Worker[MODEL]) Run(ctx context.Context) {
-	w.ctx = ctx
-	ticker := time.NewTicker(w.batchTime)
-
-	writeFn := func(context.Context) {
-		if err := w.write(ctx); err != nil {
-			// TODO(polo): should we stop the writer in case of an error ?
-			// return err
-			return
-		}
-	}
-
+func (w *Worker[MODEL]) writeLoop(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
-
-		case ms := <-w.modelsChan:
-			w.models = append(w.models, ms)
-			if w.len() >= w.batchSize {
-				writeFn(ctx)
+		case w.releasedJob <- struct{}{}:
+		case modelsHolders := <-w.jobs:
+			models := make([]MODEL, 0)
+			for _, holder := range modelsHolders {
+				models = append(models, holder.models...)
 			}
-
-		case <-ticker.C:
-			writeFn(ctx)
+			err := w.workerJob(ctx, models)
+			go func() {
+				for _, holder := range modelsHolders {
+					select {
+					case <-ctx.Done():
+						return
+					case holder.errChan <- err:
+						close(holder.errChan)
+					}
+				}
+			}()
 		}
 	}
 }
 
-func (w *Worker[MODEL]) write(ctx context.Context) error {
-	if len(w.models) == 0 {
-		return nil
-	}
+// Run should be called in a goroutine
+func (w *Worker[MODEL]) Run(ctx context.Context) {
 
-	models := make([]*MODEL, 0, w.len())
-	for _, model := range w.models {
-		models = append(models, model.models...)
-	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	err := w.workerJob(ctx, models)
+	go w.writeLoop(ctx)
 
-	for _, model := range w.models {
-		model.errChan <- err
-		close(model.errChan)
+l:
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case ch := <-w.stopChan:
+			close(ch)
+			return
+		case mh := <-w.writeChannel:
+			w.pending = append(w.pending, mh)
+		case <-w.releasedJob:
+			if len(w.pending) > 0 {
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case ch := <-w.stopChan:
+						close(ch)
+						return
+					case w.jobs <- w.pending:
```

Comment:
Should we not make a copy of the pending array ? It will be erase with the next line am I right ?

Reply:
Or maybe not, discussed IRL !